openai.go

```diff
@@ -78,6 +79,71 @@ func (s *Session) MakeRequest(ctx context.Context, endpoint string, input, outpu
 	return json.NewDecoder(resp.Body).Decode(output)
 }
 
+// Upload makes a multi-part form data upload them with
+// session's API key. Upload combines the file with the given params
+// and unmarshals the response as output.
+func (s *Session) Upload(ctx context.Context, endpoint string, file io.Reader, fileExt string, params url.Values, output interface{}) error {
+	pr, pw := io.Pipe()
+	mw := multipart.NewWriter(pw)
+	go func() {
+		err := upload(mw, file, fileExt, params)
+		pw.CloseWithError(err)
+	}()
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, pr)
+	if err != nil {
+		return err
+	}
+	if s.apiKey != "" {
+		req.Header.Set("Authorization", "Bearer "+s.apiKey)
+	}
+	if s.OrganizationID != "" {
+		req.Header.Set("OpenAI-Organization", s.OrganizationID)
+	}
+	req.Header.Set("Content-Type", mw.FormDataContentType())
+	resp, err := s.HTTPClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("error sending request: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
```

Comment:
We should probably consider splitting this into its own function.

Reply:
Yeah I thought about it, I'll push a change with the merge conflict but feel free to make further edits 👍🏼 👍🏼